cmd/kube-controller-manager/app/core.go

```diff
@@ -82,20 +82,23 @@ func startServiceController(ctx ControllerContext) (bool, error) {
 func startNodeIpamController(ctx ControllerContext) (bool, error) {
 	var clusterCIDR *net.IPNet = nil
 	var serviceCIDR *net.IPNet = nil
-	if ctx.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs {
-		var err error
-		if len(strings.TrimSpace(ctx.ComponentConfig.KubeCloudShared.ClusterCIDR)) != 0 {
-			_, clusterCIDR, err = net.ParseCIDR(ctx.ComponentConfig.KubeCloudShared.ClusterCIDR)
-			if err != nil {
-				glog.Warningf("Unsuccessful parsing of cluster CIDR %v: %v", ctx.ComponentConfig.KubeCloudShared.ClusterCIDR, err)
-			}
+
+	if !ctx.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs {
```

Comment:
I think I'm missing something. This seems to be disabling the node ipam controller on the KCM if "--allocate-node-cidrs" isn't set (https://github.com/kubernetes/kubernetes/blob/master/cmd/controller-manager/app/options/kubecloudshared.go#L57). GCE and local both explicitly set this flag to true. I also know kubeadm sets it to true based on "Networking.PodSubnet" not being null. (Not sure for which cloud providers that will be true).  My understanding is that our goal (eventual) is to have node ipam running on the KCM for all the providers except GCE. For GCE running the CCM, node ipam should be disabled on the KCM. 

I generally like this clean up but I don't think it belongs with re-enabling nodeipam in the KCM. (I would like to be able to revert one without the other) Also I don't think it meets my concerns about being able to properly control whether nodeipam is running in the KCM or CCM. (That can be handled later and is not necessarily blocking)

Reply:
> My understanding is that our goal (eventual) is to have node ipam running on the KCM for all the providers except GCE. For GCE running the CCM, node ipam should be disabled on the KCM.
Yup, this change would allow us to do that. Set `--allocate-node-cidr` if you want nodeipam on KCM (right now this is broken if you set `--cloud-provider=external` on KCM), if you are running GCE setup with CCM, make sure you set `--cloud-provider=external` and unset `--allocate-node-cidr` on KCM.

Reply:
> I generally like this clean up but I don't think it belongs with re-enabling nodeipam in the KCM. (I would like to be able to revert one without the other) Also I don't think it meets my concerns about being able to properly control whether nodeipam is running in the KCM or CCM. (That can be handled later and is not necessarily blocking) 

If you prefer, I can simplify the PR by only including [this](https://github.com/kubernetes/kubernetes/pull/63049/files#diff-2aa9cfe6a01016c1679752b0fcfc9244L332) change. But functionally it would be the same thing since nodeipam will do nothing if `--allocate-node-cidr` is not set and from GCE you would still have to unset `--allocate-node-cidr`. The nice thing about this change is that we would [explicity log that nodeipam controller was skipped](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L445).